[DebugInfo][RemoveDIs] Support maintaining DPValues in CodeGenPrepare

jmorse opened the pull request:
CodeGenPrepare needs to support the maintenence of DPValues, the non-instruction replacement for dbg.value intrinsics. This means there are a few functions we need to duplicate or replicate the functionality of:
 * fixupDbgValue for setting users of sunk addr GEPs,
 * The remains of placeDbgValues needs a DPValue implementation for sinking
 * Rollback of RAUWs needs to update DPValues
 * Rollback of instruction removal needs supporting (see github #73350)
 * A few places where we have to use iterators rather than instructions.

There are three places where we have to use the setHeadBit call on iterators to indicate which portion of debug-info records we're about to splice around. This is because CodeGenPrepare, unlike other optimisation passes, is very much concerned with which block an operation occurs in and where in the block instructions are because it's preparing things to be in a format that's good for SelectionDAG.

There isn't a large amount of test coverage for debuginfo behaviours in this pass, hence I've added some more.

jmorse committed:
[DebugInfo][RemoveDIs] Support maintaining DPValues in CodeGenPrepare

CodeGenPrepare needs to support the maintenence of DPValues, the
non-instruction replacement for dbg.value intrinsics. This means there are
a few functions we need to duplicate or replicate the functionality of:
 * fixupDbgValue for setting users of sunk addr GEPs,
 * The remains of placeDbgValues needs a DPValue implementation for sinking
 * Rollback of RAUWs needs to update DPValues
 * Rollback of instruction removal needs supporting (see github #73350)
 * A few places where we have to use iterators rather than instructions.

There are three places where we have to use the setHeadBit call on
iterators to indicate which portion of debug-info records we're about to
splice around. This is because CodeGenPrepare, unlike other optimisation
passes, is very much concerned with which block an operation occurs in and
where in the block instructions are because it's preparing things to be in
a format that's good for SelectionDAG.

There isn't a large amount of test coverage for debuginfo behaviours in
this pass, hence I've added some more.

jmorse committed:
[DebugInfo][RemoveDIs] Emulate inserting insts in dbg.value sequences

Here's a problem for the RemoveDIs project to make debug-info not be stored
in instructions -- in the following sequence:
    dbg.value(foo
    %bar = add i32 ...
    dbg.value(baz
It's possible for rare passes (only CodeGenPrepare) to remove the add
instruction, and then re-insert it back in the same place. When debug-info
is stored in instructions and there's a total order on "when" things happen
this is easy, but by moving that information out of the instruction stream
we start having to do manual maintenance.

This patch adds some utilities for re-inserting an instruction into a
sequence of DPValue objects. Someday we hope to design this away, but for
now it's necessary to support all the things you can do with dbg.values.
The two unit tests show how DPValues get shuffled around using the relevant
function calls. A follow-up patch adds instrumentation to CodeGenPrepare.

jmorse committed:
getstableloc needed in bb

jmorse committed:
Fixup reinsert to be for the "after" case.

Fix tests for insertAfter faff

jmorse commented:
Added a bonus-patch that uses `getStableDebugLoc` instead of `getStableLoc` in `BasicBlock::splitBasicBlock`.  When we spilt blocks, the default behaviour of `splitBasicBlock` is to give the resulting branch instruction the DebugLoc from the next instruction from the split position -- however if that's a dbg.value intrinsic, then we assign a source-location that's derived from a debug-intrinsic, which is wrong. It will cause differences in the line tables between -g and -gmlt.

This is covered by the test `Transforms/CodeGenPrepare/X86/cttz-ctlz.ll`: CodeGenPrepare splits the block in each function after the call to the cttz/ctlz intrinsic, which when debugify is applied is a dbg.value intrinsic. Without this change, we get different line-numbers in dbg.value mode versus RemoveDIs mode.

OCHyams commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -2234,7 +2239,9 @@ static bool despeculateCountZeros(IntrinsicInst *CountZeros,
-  BasicBlock::iterator SplitPt = ++(BasicBlock::iterator(CountZeros));
+  BasicBlock::iterator SplitPt = std::next(BasicBlock::iterator(CountZeros));
+  // Any debug-info after CountZeros should not be included.
+  SplitPt.setHeadBit(true); // coveredy by Transforms/CodeGenPrepare/X86/cttz-ctlz.ll
```
Did you intend to leave the test coverage comments in? There's are a few others scattered about too.

OCHyams commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -2829,34 +2836,46 @@ class TypePromotionTransaction {
       Instruction *PrevInst;
       BasicBlock *BB;
     } Point;
+    std::optional<DPValue::self_iterator> beforeDPValue = std::nullopt;
```
nit: beforeDPValue -> BeforeDPValue

OCHyams commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -2829,34 +2836,46 @@ class TypePromotionTransaction {
+      // of DPValues, if we ended up reinserting.
+      if (BB->IsNewDbgInfoFormat) {
+        DPMarker *DPM = BB->createMarker(Inst);
+        beforeDPValue = DPM->getReinsertionPosition();
```
Looking at the implementation of `DPMarker::getReinsertionPosition`, it looks like creating a marker just for this call is possibly a waste - I don't think any Marker-specific state is used in `getReinsertPosition`? I might be reading it wrong -- either way, I'll drop a comment on the other review.

jmorse replied:
(Will remangle with the other patch too)

OCHyams commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -2882,6 +2901,7 @@ class TypePromotionTransaction {
   };
 
   /// Set the operand of an instruction with a new value.
+  // XXX jmorse -- what about for dbg.values?
```
Leftover dev comment?

OCHyams commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -6749,7 +6775,7 @@ bool CodeGenPrepare::optimizeLoadExt(LoadInst *Load) {
     return false;
 
-  IRBuilder<> Builder(Load->getNextNode());
+  IRBuilder<> Builder(Load->getNextNonDebugInstruction());
```
:)

OCHyams commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -8095,10 +8123,14 @@ static bool optimizeBranch(BranchInst *Branch, const TargetLowering &TLI,
 bool CodeGenPrepare::optimizeInst(Instruction *I, ModifyDT &ModifiedDT) {
+  bool AnyChange = false;
+  for (DPValue &DPV : I->getDbgValueRange())
+    AnyChange |= fixupDPValue(DPV);
```
I suppose we want fixupDPValues to affect the return value "has a made change" to maintain parity with dbg.values. But is that desirable behaviour overall? Is this something we want to come back and change in the future? If so, wdyt of adding add a FIXME comment.

OCHyams commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -8359,6 +8391,30 @@ bool CodeGenPrepare::fixupDbgValue(Instruction *I) {
   return AnyChange;
 }
 
+bool CodeGenPrepare::fixupDPValue(DPValue &DPV) {
```
The debug intrinsic path asserts that it's a `DbgValueInst`. `fixupDbgValue` is only called on dbg_value (and dbg_assign)-type intrinsics, whereas `fixupDPVAlue` is called without first filtering by DPValue kind. This is fine - no action required.

OCHyams commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -8371,8 +8427,10 @@ bool CodeGenPrepare::placeDbgValues(Function &F) {
       DbgValueInst *DVI = dyn_cast<DbgValueInst>(&Insn);
-      if (!DVI)
+      if (!DVI) {
+        MadeChange |= placeDPValues(Insn, DT);
```
Shouldn't `placeDPValues` be placed before that `if (!DVI)` check, because a `DbgValueInst` can have `DPValues` attached. It's possibly an error if we have both DP- and dbg-values, so maybe it doesn't matter.

jmorse replied:
Good question -- IMO DPValues on a DbgValueInst isn't a case we should consider as it indicates that something has gone horribly wrong already (and I think there's an assertion in the basic-block converter to catch that?). Were it a DbgDeclare that'd be different as we're supporting those in parallel right now.

OCHyams commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -8421,6 +8479,65 @@ bool CodeGenPrepare::placeDbgValues(Function &F) {
+      MadeChange = true;
+      ++NumDbgValueMoved;
+    }
+  }
```
Can this be folded back into the `placeDbgValues` function using the "turn the loop into a lambda and then call for_each(DbgValues, Lambda), for_each(DPValues, Lambda)" idiom (workshopping the name still).

jmorse committed:
Review feedback

jmorse committed:
Replumb placeDbgValues to largely use a generic lambda

I can't believe that after deleting it 4 years ago I'm still working on
this function X_X

OCHyams reviewed (approved):
Looks like a couple extra commits from other reviews have  accidentally snuck in. The changes to the original patch LGTM.

jmorse committed:
Switch to new-er get-reinsertion-point method

jmorse commented:
Seeing how the commits pushed here got to be a mess, merged in 3ef98bcd46
